spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala

```diff
@@ -0,0 +1,196 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
+import org.apache.spark.sql.catalyst.plans.logical.{Call, CallArgument, CallStatement, LogicalPlan, NamedArgument, PositionalArgument}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{CatalogManager, CatalogPlugin, LookupCatalog, Procedure, ProcedureCatalog, ProcedureParameter}
+import scala.collection.Seq
+
+object ResolveProcedures extends Rule[LogicalPlan] with LookupCatalog {
+
+  protected lazy val catalogManager: CatalogManager = SparkSession.active.sessionState.catalogManager
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case CallStatement(CatalogAndIdentifier(catalog, ident), args) =>
+      val procedure = catalog.asProcedureCatalog.loadProcedure(ident)
+
+      validateParams(procedure)
+      validateMethodHandle(procedure)
+
+      Call(procedure, args = buildArgExprs(procedure, args))
+  }
+
+  private def validateParams(procedure: Procedure): Unit = {
+    // should not be any duplicate param names
+    val duplicateParamNames = procedure.parameters.groupBy(_.name).collect {
+      case (name, matchingParams) if matchingParams.length > 1 => name
+    }
+
+    if (duplicateParamNames.nonEmpty) {
+      throw new AnalysisException(s"Duplicate parameter names: ${duplicateParamNames.mkString("[", ",", "]")}")
+    }
+
+    // optional params should be at the end
+    procedure.parameters.sliding(2).foreach {
+      case Array(previousParam, currentParam) if previousParam.required && !currentParam.required =>
+        throw new AnalysisException("Optional parameters must be after required ones")
+      case _ =>
+    }
+  }
+
+  private def validateMethodHandle(procedure: Procedure): Unit = {
+    val params = procedure.parameters
+    val outputType = procedure.outputType
+
+    val methodHandle = procedure.methodHandle
+    val methodType = methodHandle.`type`
+    val methodReturnType = methodType.returnType
+
+    // method cannot accept var ags
+    if (methodHandle.isVarargsCollector) {
+      throw new AnalysisException("Method must have fixed arity")
+    }
+
+    // verify the number of params in the procedure match the number of params in the method
+    if (params.length != methodType.parameterCount) {
+      throw new AnalysisException("Method parameter count must match the number of procedure parameters")
+    }
+
+    // the MethodHandle API does not allow us to check the generic type
+    // so we only verify the return type is either void or iterable
+
+    if (outputType.nonEmpty && methodReturnType != classOf[java.lang.Iterable[_]]) {
```

Comment:
This is ugly. I'd appreciate any ideas.

Reply:
In the current approach, `methodHandle` is expected to be either void or return `java.lang.Iterable` of Spark Rows.

Reply:
Got rid of this.